File: app/src/main/java/wangdaye/com/geometricweather/weather/services/CaiYunWeatherService.java
package wangdaye.com.geometricweather.weather.services;

import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import wangdaye.com.geometricweather.BuildConfig;
import wangdaye.com.geometricweather.common.basic.models.ChineseCity;
import wangdaye.com.geometricweather.common.basic.models.Location;
import wangdaye.com.geometricweather.common.utils.LanguageUtils;
import wangdaye.com.geometricweather.common.utils.helpers.AsyncHelper;
import wangdaye.com.geometricweather.db.DatabaseHelper;
import wangdaye.com.geometricweather.weather.apis.CaiYunApi;
import wangdaye.com.geometricweather.weather.converters.CaiyunResultConverter;
import wangdaye.com.geometricweather.weather.json.caiyun.CaiYunWeatherResult;

public class CaiYunWeatherService extends WeatherService {

    private static final String TAG = "CaiYunService";

    private final CaiYunApi mApi;
    private AsyncHelper.Controller mController;

    @Inject
    public CaiYunWeatherService(CaiYunApi cyApi) {
        mApi = cyApi;
    }

    @Override
    public void requestWeather(Context context,
                               Location location, @NonNull RequestWeatherCallback callback) {
        mController = AsyncHelper.runOnIO(() -> {
            try {
                retrofit2.Response<CaiYunWeatherResult> response = mApi.getWeather(
                        BuildConfig.CAIYUN_WEATHER_KEY,
                        String.valueOf(location.getLongitude()),
                        String.valueOf(location.getLatitude()),
                        true
                ).execute();
                Log.i(TAG, "requestWeather: HTTP " + response.code() + " for "
                        + location.getLongitude() + "," + location.getLatitude());
                CaiYunWeatherResult result = response.body();
                if (result != null) {
                    WeatherResultWrapper wrapper =
                            CaiyunResultConverter.convert(context, location, result);
                    if (wrapper.result != null) {
                        Log.i(TAG, "requestWeather: conversion OK for cityId=" + location.getCityId());
                        callback.requestWeatherSuccess(
                                Location.copy(location, wrapper.result)
                        );
                    } else {
                        Log.w(TAG, "requestWeather: conversion returned null");
                        callback.requestWeatherFailed(location);
                    }
                } else {
                    Log.w(TAG, "requestWeather: body null, HTTP " + response.code()
                            + " " + response.message());
                    callback.requestWeatherFailed(location);
                }
            } catch (Exception e) {
                Log.e(TAG, "requestWeather exception", e);
                callback.requestWeatherFailed(location);
            }
        });
    }

    @NonNull
    @Override
    public List<Location> requestLocation(Context context, String query) {
        if (!LanguageUtils.isChinese(query)) {
            return new ArrayList<>();
        }

        DatabaseHelper.getInstance(context).ensureChineseCityList(context);

        List<Location> locationList = new ArrayList<>();
        List<ChineseCity> cityList = DatabaseHelper.getInstance(context).readChineseCityList(query);
        for (ChineseCity c : cityList) {
            locationList.add(c.toLocation());
        }

        return locationList;
    }

    @Override
    public void requestLocation(Context context, Location location,
                                @NonNull RequestLocationCallback callback) {
        final boolean hasGeocodeInformation = location.hasGeocodeInformation();

        mController = AsyncHelper.runOnIO(() -> {
            DatabaseHelper.getInstance(context).ensureChineseCityList(context);
            List<Location> locationList = new ArrayList<>();

            if (hasGeocodeInformation) {
                ChineseCity chineseCity = DatabaseHelper.getInstance(context).readChineseCity(
                        formatLocationString(convertChinese(location.getProvince())),
                        formatLocationString(convertChinese(location.getCity())),
                        formatLocationString(convertChinese(location.getDistrict()))
                );
                if (chineseCity != null) {
                    locationList.add(chineseCity.toLocation());
                }
            }
            if (locationList.size() > 0) {
                callback.requestLocationSuccess(location.getFormattedId(), locationList);
                return;
            }

            ChineseCity chineseCity = DatabaseHelper.getInstance(context).readChineseCity(
                    location.getLatitude(), location.getLongitude());
            if (chineseCity != null) {
                locationList.add(chineseCity.toLocation());
            }

            if (locationList.size() > 0) {
                callback.requestLocationSuccess(location.getFormattedId(), locationList);
            } else {
                callback.requestLocationFailed(location.getFormattedId());
            }
        });
    }

    @Override
    public void cancel() {
        if (mController != null) {
            mController.cancel();
        }
    }
}

File: app/src/main/java/wangdaye/com/geometricweather/weather/services/CaiYunWeatherService.kt
package wangdaye.com.geometricweather.weather.services

import android.content.Context
import kotlinx.coroutines.isActive
import wangdaye.com.geometricweather.BuildConfig
import wangdaye.com.geometricweather.common.basic.models.Location
import wangdaye.com.geometricweather.common.utils.LanguageUtils
import wangdaye.com.geometricweather.db.DatabaseHelper
import wangdaye.com.geometricweather.weather.apis.CaiYunApi
import wangdaye.com.geometricweather.weather.converters.CaiyunResultConverter
import javax.inject.Inject

/**
 * CaiYun (彩云天气) service, on the official v2.6 token endpoint.
 *
 * Weather is one call, keyed by "{lon},{lat}" in the URL path — longitude first. Location search
 * does not go to the network at all: it reads the bundled Chinese city table out of Room, which is
 * also how a current position is matched to a city.
 */
class CaiYunWeatherService @Inject constructor(
    private val api: CaiYunApi
) : WeatherService() {

    private val requests = RequestScope()

    override fun requestWeather(
        context: Context,
        location: Location,
        callback: RequestWeatherCallback
    ) {
        requests.launch {
            val result = requests.execute(
                api.getWeather(
                    BuildConfig.CAIYUN_WEATHER_KEY,
                    location.longitude.toString(),
                    location.latitude.toString(),
                    true
                )
            )
            val weather = result?.let {
                CaiyunResultConverter.convert(context, location, it).result
            }

            // Nothing below this point may reach the caller once cancel() has been called.
            if (!isActive) {
                return@launch
            }
            if (weather != null) {
                callback.requestWeatherSuccess(Location.copy(location, weather))
            } else {
                callback.requestWeatherFailed(location)
            }
        }
    }

    override fun requestLocation(context: Context, query: String): List<Location> {
        if (!LanguageUtils.isChinese(query)) {
            return emptyList()
        }
        val database = DatabaseHelper.getInstance(context)
        database.ensureChineseCityList(context)
        return database.readChineseCityList(query).map { it.toLocation() }
    }

    override fun requestLocation(
        context: Context,
        location: Location,
        callback: RequestLocationCallback
    ) {
        val hasGeocodeInformation = location.hasGeocodeInformation()

        requests.launch {
            val database = DatabaseHelper.getInstance(context)
            database.ensureChineseCityList(context)

            // Prefer the reverse-geocoded names; fall back to matching the raw coordinates.
            val city = if (hasGeocodeInformation) {
                database.readChineseCity(
                    formatLocationString(convertChinese(location.province)),
                    formatLocationString(convertChinese(location.city)),
                    formatLocationString(convertChinese(location.district))
                )
            } else {
                null
            } ?: database.readChineseCity(location.latitude, location.longitude)

            if (!isActive) {
                return@launch
            }
            if (city != null) {
                callback.requestLocationSuccess(location.formattedId, listOf(city.toLocation()))
            } else {
                callback.requestLocationFailed(location.formattedId)
            }
        }
    }

    override fun cancel() = requests.cancel()
}
